src/common.cpp

```diff
@@ -2299,34 +2299,40 @@ uint32_t ff7_get_inserted_cd(void) {
 	int ret = 1;
 	int requiredCD;
 	int insertedCD;
+	DWORD* insertedCDAddress;
 
 	switch (version)
 	{
 	case VERSION_FF7_102_US:
 		requiredCD = *(uint8_t*)(0xDC0BDC);
-		insertedCD = *(DWORD*)(0x9A0538);
+		insertedCDAddress = (DWORD*)(0x9A0538);
 		break;
 	case VERSION_FF7_102_FR:
 		requiredCD = *(uint8_t*)(0xF3A91C);
-		insertedCD = *(DWORD*)(0x9A2328);
+		insertedCDAddress = (DWORD*)(0x9A2328);
 		break;
 	case VERSION_FF7_102_DE:
 		requiredCD = *(uint8_t*)(0xF3990C);
-		insertedCD = *(DWORD*)(0x9A12F8);
+		insertedCDAddress = (DWORD*)(0x9A12F8);
 		break;
 	case VERSION_FF7_102_SP:
 		requiredCD = *(uint8_t*)(0xF3B3EC);
-		insertedCD = *(DWORD*)(0x9A2D88);
+		insertedCDAddress = (DWORD*)(0x9A2D88);
 		break;
 	}
+	insertedCD = *insertedCDAddress;
 
-	if(steam_edition || enable_steam_achievements)
-		if(insertedCD != requiredCD)
-			g_FF7SteamAchievements.unlockGameProgressAchievement();
+	if(steam_edition || enable_steam_achievements){
+		if (trace_all || trace_achievement)
+			ffnx_trace("inserted CD: %d, requiredCD: %d\n", insertedCD, requiredCD);
+
+		if(requiredCD == insertedCD + 1)
+			g_FF7SteamAchievements->unlockGameProgressAchievement();
+	}
 
 	if (requiredCD > 0 && requiredCD <= 3) ret = requiredCD;
 
-	insertedCD = ret;
+	*insertedCDAddress = ret;
```

Comment:
I get what you're trying to do it here but I'd like to see the patch done differently.

Move the CD declaration variable in the `ff7_externals` and use it to read/write into the location. Also use a proper name for the variable like `insertedCD`, even if it's a pointer to an address, like most of the declarations are in this code anyway. No need to suffix it with `Address`.

Thanks :)

Reply:
Now, it should be fine